MRN-Code/coinstac
# Hard Links 'n Pullin' Comps

* **Problem:**
   1. Pipeline needs hard links to user data files.
   1. Before a pipeline can start, users need to download the appropriate pipeline computations to their machines.
* **Solution:**
   1. A function `startPipeline` was created in `coinstac-client-core` that will create hard links to all paths in the `filesArray` argument.
   1. Pipeline, consortia and computation subscriptions were moved to top level component in order to receive, and act on, activePipelineId changes in `consortia`.
   1. Computations will now download on a client's machine when one of the following is true:
       1. A user joins a consortium with an existing `activePipelineId`
       1. An `activePipelineId` is set on a consortium to which a user already belongs
   ~1. A new navigation item - **Docker** - was added to display the stream output as Docker downloads.~
   1. Docker output is redirected to Computation navigation item beneath the output's respective computation list item.
   1. A *Download All* button has been added to the computations page to allow a user to download all computations.
   1. Each computation now has a *Download* button which will allow users to individually download docker images.
   1. Docker remove functionality has been added for images the system detects locally.
* **Tests:**
   1. Run `npm run test-setup` from the `packages/coinstac-api-server` directory
   1. Start the application and login as `author`.
   1. Navigate to the **Consortia** nav item and leave, then join, **Test Consortium 1**.
   1. The application should start downloading the pipeline computations.
   1. Open a new instance of the application and login as `test`.
   1. Navigate to the **Consortia** item and view the pipeline details of **Test Consortium 2**.
   1. In the instance logged-in by author, add an active pipeline to **Test Consortium 2**.
   1. Docker output should display in the Computation tab on the `test` instance indicating that the requested computations are already up-to-date.
